dialog/file_test.go

```diff
@@ -505,13 +505,86 @@ func TestView(t *testing.T) {
 	assert.Equal(t, "Dismiss", dismiss.Text)
 }
 
+func TestSetView(t *testing.T) {
+	win := test.NewWindow(widget.NewLabel("Content"))
+
+	fyne.CurrentApp().Preferences().SetInt(viewLayoutKey, int(defaultView))
+
+	dlg := NewFileOpen(func(reader fyne.URIReadCloser, err error) {
+		assert.Nil(t, err)
+		assert.Nil(t, reader)
+	}, win)
+
+	dlg.SetConfirmText("Yes")
+	dlg.SetDismissText("Dismiss")
+
+	// set view to list
+	dlg.SetView(ListView)
+
+	dlg.Show()
+
+	popup := win.Canvas().Overlays().Top().(*widget.PopUp)
+	defer win.Canvas().Overlays().Remove(popup)
+	assert.NotNil(t, popup)
+
+	ui := popup.Content.(*fyne.Container)
+	toggleViewButton := ui.Objects[1].(*fyne.Container).Objects[0].(*fyne.Container).Objects[1].(*widget.Button)
+	panel := ui.Objects[0].(*container.Split).Trailing.(*fyne.Container).Objects[1].(*container.Scroll).Content.(*fyne.Container).Objects[0]
+
+	// view should be a list
+	_, isList := panel.(*widget.List)
+	assert.True(t, isList)
+	// toggleViewButton should reflect to what it will do (change to a grid view).
+	assert.Equal(t, "", toggleViewButton.Text)
+	assert.Equal(t, theme.GridIcon(), toggleViewButton.Icon)
+
+	confirm := ui.Objects[2].(*fyne.Container).Objects[0].(*fyne.Container).Objects[1].(*widget.Button)
+	assert.Equal(t, "Yes", confirm.Text)
+	dismiss := ui.Objects[2].(*fyne.Container).Objects[0].(*fyne.Container).Objects[0].(*widget.Button)
+	assert.Equal(t, "Dismiss", dismiss.Text)
+}
+
+func TestSetViewPreferences(t *testing.T) {
+	win := test.NewWindow(widget.NewLabel("Content"))
+
+	prefs := fyne.CurrentApp().Preferences()
+
+	// set user-saved viewLayout to GridView
+	prefs.SetInt(viewLayoutKey, int(GridView))
+
+	dlg := NewFileOpen(func(reader fyne.URIReadCloser, err error) {
+		assert.Nil(t, err)
+		assert.Nil(t, reader)
+	}, win)
+
+	// set default view to be ListView
+	dlg.SetView(ListView)
+
+	dlg.Show()
+
+	popup := win.Canvas().Overlays().Top().(*widget.PopUp)
+	defer win.Canvas().Overlays().Remove(popup)
+	assert.NotNil(t, popup)
+
+	ui := popup.Content.(*fyne.Container)
+	toggleViewButton := ui.Objects[1].(*fyne.Container).Objects[0].(*fyne.Container).Objects[1].(*widget.Button)
+	panel := ui.Objects[0].(*container.Split).Trailing.(*fyne.Container).Objects[1].(*container.Scroll).Content.(*fyne.Container).Objects[0]
+
+	// check that preference setting overrules configured default view
+	_, isGrid := panel.(*widget.GridWrap)
+	assert.True(t, isGrid)
+	// toggleViewButton should reflect to what it will do (change to a list view).
+	assert.Equal(t, "", toggleViewButton.Text)
+	assert.Equal(t, theme.ListIcon(), toggleViewButton.Icon)
+}
+
 func TestViewPreferences(t *testing.T) {
 	win := test.NewWindow(widget.NewLabel("Content"))
 
 	prefs := fyne.CurrentApp().Preferences()
 
 	// set viewLayout to an invalid value to verify that this situation is handled properly
-	prefs.SetInt(viewLayoutKey, -1)
```

Comment:
I am not sure when this would be the case.
I checked that `Preferences().Int()` returns 0 if the key is not found, so I think the current code should be safe and backward compatible.

Reply:
As the comment above says, it is setting an invalid value which is a good test case.
Preferences could be set by an external force and so we should be safe and handle unexpected values.

Reply:
Thanks, you are right. I added code to handle invalid values as well.